### DISCUSSION TITLE
(Slow) Vanilla Visidata in the browser

### DISCUSSION THREAD

**Discussion by asteroidb612** [saulpw/visidata#2654] [Show and tell]
I have been tinkering with an easier way to share visidata with folks who aren't as terminal-savvy, and here is my recent milestone: https://asteroidb612.github.io/wisidata/

![Screenshot 2024-12-26 at 11 43 10 AM](https://github.com/user-attachments/assets/1630f9e8-d837-42f3-9c16-6c1c5d95e041)

This is the full visidata application running client side in a browser thanks to some cool work by the [WebVM](https://webvm.io/) project.

It took some fiddling with their templates to get it to work, and it still is unusably slow, but I am excited at the prospect of sending someone a link to a [replay'd visidata session ](https://www.visidata.org/docs/save-restore/) which they can step through and explore for themselves, without any installation work. 

To that end, I need some help knowing how to improve my demo's speed and reliability. There may be tricks on the webvm side or the visidata side that would help me make a better demo - I am new to both! 

If you know good ways to see into visidata, or some features to turn off that might boost performance, or have any other thoughts to improve this project, please share.

**Comment by midichef** [comment]
Wow, interesting concept. I hope you can get it working. That would be great for sharing visidata demos.

On my weak machine, visidata in WebVM was too slow to evaluate. Running `vd --version` took > 20 seconds. Running vd interactively took several minutes to eventually draw a 1 row sheet. So the speedups we need are in the range of 10-100x. I don't have any suggestions that would give that kind of boost.

But after you get it faster, and are ready for lesser speedups, try adjusting `timeouts_before_idle` to be `1` instead of `10`:
https://github.com/saulpw/visidata/blob/efbe84e6d89ab1442a45f4a6a651b4791f61d175/visidata/mainloop.py#L14
By default, calculations for columns are run 10 times, about 100ms apart, before the main loop goes idle to wait for a keypress. You can see the 10 calculations if you put `import time` in your .visidatarc, and then create a column with the time via `=` `time.time()`.

Another optimization, that is much less significant, would be to turn off the `undo` option by running vd as `vd --undo=False`. Many commands save a copy of every row in the sheet, to be later restored by `undo-last.` (Note that undo is already disabled in batch mode (`-b`) when interactive (`-i`) is not present.) Turning off `undo` would save a small amount of memory and compute.

**Reply by asteroidb612** [reply]
Thanks for the tips!

I'm seeing ~100x slowdowns on my machine as well. 

The fact that these are all before anything's even loaded makes me think there's more to be done on the webvm side than the visidata side. Their expectation seems to be 10x slowdowns, so maybe this can get to a workable state on a fast machine. 

I did find this thread helpful to know some debugging tips on the vd side though: https://github.com/saulpw/visidata/issues/558
